chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -423,55 +423,36 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
         Ok(())
     }
 
-    fn check_block_reward_timelock(
-        &self,
-        block: &Block,
-        timelock: &OutputTimeLock,
-    ) -> Result<(), CheckBlockError> {
-        match timelock {
-            OutputTimeLock::ForBlockCount(c) => {
-                let cs: i64 = (*c)
-                    .try_into()
-                    .map_err(|_| {
-                        CheckBlockError::InvalidBlockRewardMaturityDistanceValue(block.get_id(), *c)
-                    })
-                    .log_err()?;
-                let given = BlockDistance::new(cs);
-                let required = block.consensus_data().reward_maturity_distance(self.chain_config);
-                ensure!(
-                    given >= required,
-                    CheckBlockError::InvalidBlockRewardMaturityDistance(
-                        block.get_id(),
-                        given,
-                        required
-                    )
-                );
-                Ok(())
-            }
-            OutputTimeLock::UntilHeight(_)
-            | OutputTimeLock::UntilTime(_)
-            | OutputTimeLock::ForSeconds(_) => Err(
-                CheckBlockError::InvalidBlockRewardMaturityTimelockType(block.get_id()),
-            ),
-        }
-    }
-
     fn check_block_reward_maturity_settings(&self, block: &Block) -> Result<(), CheckBlockError> {
-        block.block_reward().outputs().iter().try_for_each(|output| {
-            match output {
-                TxOutput::LockThenTransfer(_, _, tl) => self.check_block_reward_timelock(block, tl),
-                TxOutput::ProduceBlockFromStake(_, _) => {
-                    // The output can be reused in block reward right away
-                    Ok(())
+        block
+            .block_reward()
+            .outputs()
+            .iter()
+            .enumerate()
+            .try_for_each(|(index, output)| {
+                match output {
+                    TxOutput::LockThenTransfer(_, _, tl) => {
+                        let outpoint = OutPoint::new(block.get_id().into(), index as u32);
+                        let required =
+                            block.consensus_data().reward_maturity_distance(self.chain_config);
+                        tx_verifier::timelock_check::check_output_maturity_setting(
+                            tl, required, outpoint,
+                        )
+                        .map_err(CheckBlockError::BlockRewardMaturityError)
+                    }
```

Comment:
I'm confused... are we allowing putting a LockThenTransfer in a BlockReward output?

Reply:
for PoW yes but not for PoS. The checks are in `input_output_policy.rs`
This particular function performs isolated output timelock  checks only.

Reply:
what I did is reused a function from tx_verifier

Reply:
Honestly, I don't like this... I would prefer to have a separate function for each than this unexplainable mess...

If you want to check for maturity, then check for maturity. What bothers me here is that the function claims that it checks for the maturity, then we kind of return an error for other types too.

In my head, the right thing to do is: You match for consensus type, then you call a separate function that does the checking for that type alone, where one is PoW, and checks LockThenTransfer, and the other is PoS, and allows ProduceBlockFromStake with no limitations. This way, if we add new consensus types in the future (say, PoSv2), we can simply add another function with a separate check.

Mixing and matching like this isn't really correct and creates attack vectors in the future when we get more and more consensus types because the number of combinations can grow exponentially.

Reply:
ok I see